### ISSUE TITLE
Build component props from query/state?

### ISSUE THREAD

**Issue by hagabaka** [kitbagjs/router#340]
When defining a route, currently you can [provide component props based on the route params](https://kitbag-router.netlify.app/core-concepts/component-props.html):
<blockquote>

```ts
const user = createRoute({
  props: (params) => {
    return { foo: params.bar }
  }
})
```

</blockquote>

But the props getter doesn't have access to route query and state, even though as they can be passed when navigating to a route like params. One reason why I would like to to use state instead of a param is to pass a complex value for a component prop, without having a long URL.

**Comment by pleek91** [comment]
@hagabaka great suggestion. Let me look into adding some more information about the route itself into the [callback's context](https://kitbag-router.netlify.app/core-concepts/component-props.html#context)

**Comment by pleek91** [comment]
@hagabaka take a look at the linked PR and let me know what you think. I believe that should handle your use case and its much more flexible.

**Comment by hagabaka** [comment]
> @hagabaka take a look at the linked PR and let me know what you think. I believe that should handle your use case and its much more flexible.

The pull request looks good, however it seems like only String states are passed to the props function, and not other types. Here are two test cases showing this (could be added to src/components/routerView.browser.spec.ts):
```ts
test('Binds props and attrs from route state (string)', async () => {
  const route = createRoute({
    name: 'propsFromStringState',
    path: '/propsFromStringState',
    component: echo,
    state: { value: String },
    props: ({ state: { value = 'missing state' } }) => ({ value }),
  })

  const router = createRouter([route], {
    initialUrl: '/',
  })

  await router.start()

  const root = {
    template: '<RouterView/>',
  }

  const app = mount(root, {
    global: {
      plugins: [router],
    },
  })

  await router.push('/propsFromStringState', { state: { value: 'hello' } })
  expect(app.html()).toBe('hello')
})

test('Binds props and attrs from route state (number)', async () => {
  const route = createRoute({
    name: 'propsFromNumberState',
    path: '/propsFromNumberState',
    component: echo,
    state: { value: String },
    props: ({ state: { value = 'missing state' } }) => ({ value }),
  })

  const router = createRouter([route], {
    initialUrl: '/',
  })

  await router.start()

  const root = {
    template: '<RouterView/>',
  }

  const app = mount(root, {
    global: {
      plugins: [router],
    },
  })

  await router.push('/propsFromNumberState', { state: { value: 42 } })
  expect(app.html()).toBe('42')
})
```
```
   ✓ Binds props and attrs from route state (string)
   × Binds props and attrs from route state (number)
 FAIL  src/components/routerView.browser.spec.ts > Binds props and attrs from route state (number)
AssertionError: expected 'missing state' to be '42' // Object.is equality

Expected: "42"
Received: "missing state"
```

**Comment by pleek91** [comment]
@hagabaka thanks for the tests! I'll look into this

**Comment by stackoverfloweth** [comment]
Technically you could continue to route by path, but the options are not type safe and expected to be in "raw" format. Which in this case means the values for the state should be strings.

```ts
await router.push('/propsFromNumberState', { state: { value: '42' } })
```

Note the router will still use the param getter to convert that raw value into a number when you access it.

**Comment by pleek91** [comment]
Closing this issue based on the props callback changes released in 0.12.0. Please open another issue for any other requests you have or bugs you find. Thank you for contributing!
